File: backend/Models/Post.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace backend.Models
{
    public class Post
    {
        [Key]
        public long PostId { get; set; }

        [Required]
        [StringLength(2000, ErrorMessage = "Post content cannot exceed 2000 characters")]
        public string? Content { get; set; }

        public string? ImageUrl { get; set; }

        [StringLength(500, ErrorMessage = "Image alt text cannot exceed 500 characters")]
        public string? ImageAltText { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        // Privacy settings
        public PostPrivacy Privacy { get; set; } = PostPrivacy.Public;

        // Relationships
        [Required]
        public long UserId { get; set; }
        public User? User { get; set; }

        // Optional trip reference - posts can be about specific trips
        public long? TripId { get; set; }
        public Trip? Trip { get; set; }

        // Navigation properties for interactions
        public ICollection<PostLike> Likes { get; set; } = new List<PostLike>();
        public ICollection<PostComment> Comments { get; set; } = new List<PostComment>();

        // Computed properties for counts
        [NotMapped]
        public int LikeCount => Likes?.Count ?? 0;

        [NotMapped]
        public int CommentCount => Comments?.Count ?? 0;
    }

    public enum PostPrivacy
    {
        Public = 0,
        FollowersOnly = 1,
        Private = 2
    }

    public class PostLike
    {
        [Key]
        public long PostLikeId { get; set; }

        [Required]
        public long PostId { get; set; }

        [JsonIgnore] // Prevent circular reference
        public Post? Post { get; set; }

        [Required]
        public long UserId { get; set; }
        public User? User { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    }

    public class PostComment
    {
        [Key]
        public long PostCommentId { get; set; }

        [Required]
        [StringLength(1000, ErrorMessage = "Comment cannot exceed 1000 characters")]
        public string? Content { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        // Relationships
        [Required]
        public long PostId { get; set; }

        [JsonIgnore] // Prevent circular reference
        public Post? Post { get; set; }

        [Required]
        public long UserId { get; set; }
        public User? User { get; set; }
    }
}

File: frontend/src/models/post.ts
// Removed unused imports - user and trip info is now flattened in DTOs

export interface Post {
	postId: number;
	content: string;
	imageUrl?: string;
	imageAltText?: string;
	createdAt: string;
	updatedAt?: string;
	privacy: PostPrivacy;
	userId: number;
	// Flattened user info (from DTO)
	userFirstName?: string;
	userLastName?: string;
	userProfilePicture?: string;
	// Flattened trip info (from DTO)
	tripId?: number;
	tripName?: string;
	tripDestination?: string;
	// Simple counts
	likeCount: number;
	commentCount: number;
	// Simple lists without circular references
	likes: PostLike[];
	comments: PostComment[];
}

export const PostPrivacy = {
	Public: 0,
	FollowersOnly: 1,
	Private: 2,
} as const;

export type PostPrivacy = (typeof PostPrivacy)[keyof typeof PostPrivacy];

export interface PostLike {
	postLikeId: number;
	postId: number;
	userId: number;
	// Flattened user info (from DTO)
	userFirstName?: string;
	userLastName?: string;
	userProfilePicture?: string;
	createdAt: string;
}

export interface PostComment {
	postCommentId: number;
	content: string;
	createdAt: string;
	updatedAt?: string;
	postId: number;
	userId: number;
	// Flattened user info (from DTO)
	userFirstName?: string;
	userLastName?: string;
	userProfilePicture?: string;
}

export interface CreatePostRequest {
	content: string;
	image?: File;
	imageAltText?: string;
	tripId?: number;
	privacy?: PostPrivacy;
}

export interface UpdatePostRequest {
	content?: string;
	image?: File;
	imageAltText?: string;
	tripId?: number;
	privacy?: PostPrivacy;
	removeImage?: boolean;
}

export interface CreateCommentRequest {
	content: string;
}

export interface UpdateCommentRequest {
	content: string;
}

export interface PostFeedParams {
	page?: number;
	pageSize?: number;
}
